jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ThreadIdPool.java

```diff
@@ -0,0 +1,252 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util.thread;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.eclipse.jetty.util.MemoryUtils;
+import org.eclipse.jetty.util.ProcessorUtils;
+import org.eclipse.jetty.util.TypeUtil;
+import org.eclipse.jetty.util.component.Dumpable;
+import org.eclipse.jetty.util.component.DumpableCollection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A fixed sized pool of items that uses ThreadId to avoid contention.
+ * This class can be used, instead of a {@link ThreadLocal}, when pooling items
+ * that are expensive to create, but only used briefly in the scope of a single thread.
+ * It is safe to use with {@link org.eclipse.jetty.util.VirtualThreads}, as unlike a {@link ThreadLocal} pool,
+ * the number of items is limited.
+ * <p>This is a light-weight version of {@link org.eclipse.jetty.util.ConcurrentPool} that is best used
+ * when items do not reserve an index in the pool even when acquired.
+ * @see org.eclipse.jetty.util.ConcurrentPool
+ */
+public class ThreadIdPool<E> implements Dumpable
+{
+    private static final Logger LOG = LoggerFactory.getLogger(ThreadIdPool.class);
+
+    // How far the entries in the AtomicReferenceArray are spread apart to avoid false sharing;
+    // use getIntegersPerCacheLine() instead of getLongsPerCacheLine() b/c references could be compressed.
+    private static final int SPREAD_FACTOR = MemoryUtils.getIntegersPerCacheLine();
+
+    private final int _capacity;
+    private final AtomicReferenceArray<E> _items;
+
+    public ThreadIdPool()
+    {
+        this(-1);
+    }
+
+    public ThreadIdPool(int capacity)
+    {
+        _capacity = TypeUtil.ceilToNextPowerOfTwo(capacity);
+        _items = new AtomicReferenceArray<>((calcCapacity(_capacity) + 1) * SPREAD_FACTOR);
```

Comment:
I think we want to store the "normalized" capacity after the heuristics done in `calcCapacity()`, as the parameter can be negative.

Reply:
I've moved it to be in the calcCapacity, so if an explicit capacity is set, it is used.
Also doing similar logic in ReservedThreadExecutor

Reply:
@gregw the `_capacity` must be whatever `calcCapacity()` computes. The current code is backwards: I pass -1 to expect the number of cores or similar heuristic, TypeUtil converts it to 1 (why? it should just throw), and now we are using the wrong capacity.